### PR TITLE
Redesign Magnet Class to be built from YAML config file 

### PR DESCRIPTION
**Pull request by MattKing06**
## Summary of Changes

- Added Device Base Class
  - Creates `control_information` and `metadata`  dict attributes from the dictionary given to `__init__` 
  - Handles all callback related functionality that was in `Magnet
- Added Magnet Class
  - Child class of `Device`
  - Creates attributes and assigns them values from the dictionary given to `__init__`
- Added Reader util
  - Simple function to read YAML file and create magnet
  - If no specific magnet name is given to the reader, it will create all magnets in the yaml file and return a `dict {<name> : Magnet}`
- Added `gun.yaml` file
  - Serves as a POC for the yaml layout
  - Only contains normalconducting gun solenoids 
 
## Tests

- Unit tests have been added for `reader.py`, `magnet.py`, and `device.py` files
- You can test manually and play around with the changes using the script below:

```python
import os
from lcls_tools.common.devices.magnet.reader import create_magnet

YAML_LOCATION = './lcls_tools/normalconducting/device/config/magnet/'
YAML_FILE = 'gun.yaml'

magnet_name = 'SOL1B'
magnet_object = create_magnet(os.path.join(YAML_LOCATION, YAML_FILE), magnet_name)
print(magnet_object.name)

# Look in lcls_tools/common/devices/magnet/magnet.py to see more functions.
```

## Notes
The location of the yaml files will most likely change after #92 
This was not testing against EPICS and should be tested at the earliest opportunity

### CODE REVIEW

File: tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py [hunk 1]
```diff
@@ -1,105 +1,215 @@
...
+    @patch("epics.PV.get", new_callable=Mock)
+    def test_get_ctrl_calls_pv_get(self, pv_get_mock):
+        pv_get_mock.return_value = "Ready"
+        value = self.magnet.ctrl
+        self.assertEqual(value, "Ready")
+        pv_get_mock.assert_called_once_with(as_string=True)
+
+    @patch("epics.PV.put", new_callable=Mock)
+    @patch(
+        "lcls_tools.common.devices.magnet.magnet.YAMLMagnet.ctrl",
```
Comment (nneveu) [inline tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py:159]: Change from YAMLMagnet to magnet.

File: tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py [hunk 2]
```diff
@@ -1,105 +1,215 @@
...
+        for item in [
+            "bctrl",
+            "bact",
+            "ctrl_value",
+            "length",
+            "b_tolerance",
+        ]:
+            self.assertTrue(
+                hasattr(self.magnet, "bctrl"),
+                msg=f"expected magnet to have attribute {item}",
```
Comment (nneveu) [inline tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py:54]: change bctrl to item

File: tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py [hunk 3]
```diff
@@ -1,105 +1,215 @@
...
+            )
+        # Assert that magnet has public properties
+        for item in [
+            "bctrl",
+            "bact",
+            "ctrl_value",
+            "length",
+            "b_tolerance",
+        ]:
```
Comment (nneveu) [inline tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py:51]: Maybe use the yaml template instead?

File: tests/unit_tests/lcls_tools/common/devices/test_device.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        )
+        self.device.add_callback_to_pv(
+            "bact",
+            third_callback,
+        )
+        index = self.device._get_callback_index(
+            mock_pv,
+            second_callback,
+        )
+        self.assertEqual(index, 2)
```
Comment (nneveu) [inline tests/unit_tests/lcls_tools/common/devices/test_device.py:147]: comments

**Review by nneveu — CHANGES_REQUESTED** [review]

File: tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py (hunk 2 above)
Comment (MattKing06) [reply tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py:54]: Done

File: tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py (hunk 1 above)
Comment (MattKing06) [reply tests/unit_tests/lcls_tools/common/devices/magnet/test_magnet.py]: Done :smile:

File: tests/unit_tests/lcls_tools/common/devices/test_device.py (hunk 4 above)
Comment (MattKing06) [reply tests/unit_tests/lcls_tools/common/devices/test_device.py:151]: Added some comments for the more opaque tests in test_device.py

Comment (nneveu) [reply tests/unit_tests/lcls_tools/common/devices/test_device.py:151]: Thanks!

**Review by nneveu — APPROVED** [review]

**Pull request merged**